onnxruntime/core/framework/random_generator.h

```diff
@@ -0,0 +1,34 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+#pragma once
+
+#include <atomic>
+#include <stdint.h>
+#include <utility>
+
+namespace onnxruntime {
+
+class PhiloxGenerator {
+public:
+  PhiloxGenerator() : seed_(0), offset_(0) {}
+  PhiloxGenerator(uint64_t seed) : seed_(seed), offset_(0) {}
+
+  void SetSeed(uint64_t seed) {
+    seed_ = seed;
+    offset_.store(0);
+  }
+
+  std::pair<uint64_t, uint64_t> GetPhiloxSeeds(uint64_t count) {
+    uint64_t offset = offset_.fetch_add(count);
+    return std::pair<uint64_t, uint64_t>(seed_, offset);
+  }
+
+  static PhiloxGenerator& Default();
+
+ private:
+  uint64_t seed_;
```

Comment:
why is seed_ not atomic? perhaps it would be good to document how thread-safe this class is, i.e., what methods can be called concurrently?

Reply:
also what's a philox? :)

Reply:
Good questions.  I'll add documentation.

[Philox](https://docs.scipy.org/doc/numpy/reference/random/bit_generators/philox.html) is a particular random number generator technique.  CUDA (curand specifically) uses it, which is how we implement the CUDA Dropout kernel.

Reply:
thanks for adding documentation. regarding the thread-safety, i wonder if both seed_ and offset_ need to be updated together atomically? consider concurrent calls to SetSeed() and NextPhiloxSeeds():

SetSeed() | NextPhiloxSeeds()
-|-
seed_ = seed; ||
|| uint64_t offset = offset_.fetch_add(count);
offset_.store(0); ||
|| return std::pair<uint64_t, uint64_t>(seed_, offset);

could we get an inconsistent pair?

Reply:
Yes we can.  Generally speaking, seed is set at start of process and then fixed after that.  But you're right this isn't strictly guaranteed.  I was trying to avoid using mutex but, I'll use a mutex just for safety.

Reply:
might be possible to do some fancier stuff like packing two values into a single atomic one, but i suppose until there's a demonstrated need, a mutex is just fine.
nit: now maybe we don't need to #include <atomic> anymore